### PR TITLE
feat: Handle GVK validation on CRP Resource Selector

### PR DESCRIPTION
**Pull request by britaniar**
### Description of your changes

Fixes #

I have: added a GVK and cluster scope check to crp validator and corresponding unit test cases.

- [x] Run `make reviewable` to ensure this PR is ready for review.

### CODE REVIEW

File: pkg/utils/validator/clusterresourceplacement_test.go [hunk 1]
```diff
@@ -199,9 +274,10 @@ func TestValidateClusterResourcePlacementAlpha(t *testing.T) {
```
Comment (zhiying-lin) [inline pkg/utils/validator/clusterresourceplacement_test.go:274]: you can revert the v1alpha1 changes and you don't need to enhance the deprecating api

File: test/e2e/placement_selecting_resources_test.go [hunk 2]
```diff
@@ -609,11 +609,10 @@ var _ = Describe("validating CRP when deleting resources in a matching namespace
 	})
 })
 
-// TODO should be blocked by the validation webhook
 var _ = Describe("validating CRP when resource selector is not valid", Ordered, func() {
 	crpName := fmt.Sprintf(crpNameTemplate, GinkgoParallelProcess())
 
-	BeforeAll(func() {
+	It("should be denied by webhook", func() {
```
Comment (zhiying-lin) [inline test/e2e/placement_selecting_resources_test.go:615]: add an e2e test if the resource selector is selecting namespace scope resource.

File: test/e2e/placement_selecting_resources_test.go [hunk 3]
```diff
@@ -634,52 +633,7 @@ var _ = Describe("validating CRP when resource selector is not valid", Ordered,
...
-				Name: crpName,
-			},
-		}
-		Expect(hubClient.Delete(ctx, crp)).To(Succeed(), "Failed to delete CRP %s", crpName)
-	})
-
-	It("should remove controller finalizers from CRP", func() {
-		finalizerRemovedActual := allFinalizersExceptForCustomDeletionBlockerRemovedFromCRPActual()
-		Eventually(finalizerRemovedActual, eventuallyDuration, eventuallyInterval).Should(Succeed(), "Failed to remove controller finalizers from CRP %s", crpName)
+		Eventually(hubClient.Create(ctx, crp), eventuallyDuration, eventuallyInterval).ShouldNot(Succeed(), "Validation webhook did catch invalid cluster resource in crp", crpName)
```
Comment (zhiying-lin) [inline test/e2e/placement_selecting_resources_test.go:636]: you don't need the eventually. The create should fail immediately + please check the error string,

eg,

err := hubClient.Create(ctx, crp)
		var statusErr *k8sErrors.StatusError
		Expect(errors.As(err, &statusErr)).To(BeTrue(), fmt.Sprintf("Create CRP call produced error %s. Error type wanted is %s.", reflect.TypeOf(err), reflect.TypeOf(&k8sErrors.StatusError{})))
		Expect(statusErr.Status().Message).Should(ContainSubstring(fmt.Sprintf("the labelSelector and name fields are mutually exclusive in selector %+v", selector[0])))

can you move this test to the webhook_test.go?

**Review by zhiying-lin — COMMENTED** [review]
please revert any v1alpha1 changes, thanks

File: pkg/utils/common.go [hunk 4]
```diff
@@ -128,6 +128,18 @@ var (
 		Resource: placementv1beta1.ClusterResourcePlacementResource,
 	}
 
+	ClusterRoleGVK = schema.GroupVersionKind{
```
Comment (zhiying-lin) [inline pkg/utils/common.go:131]: can you move these definition to your test file or test/util package? these are not fleet related resources.

Prefer not to mix the test code with production code.

File: pkg/controllers/rollout/suite_test.go [hunk 5]
```diff
@@ -91,6 +94,20 @@ var _ = BeforeSuite(func() {
...
+	validator.ResourceInformer = &informer.FakeManager{
+		APIResources: map[schema.GroupVersionKind]bool{
+			{
+				Group:   "",
+				Version: "v1",
+				Kind:    "Namespace",
+			}: true,
+		},
+		IsClusterScopedResource: true,
+	}
```
Comment (zhiying-lin) [inline pkg/controllers/rollout/suite_test.go:109]: just curious why we need these. if i member it correctly, rollout controller does not call webhook?

Comment (britaniar) [reply pkg/controllers/rollout/suite_test.go:109]: The rollout controller integration test [creates a crp](https://github.com/Azure/fleet/blob/3c13e7d99cf58f6a7e34a55ec559424fceb993b0/pkg/controllers/rollout/controller_integration_test.go#L66)

File: pkg/controllers/rollout/manifests/test_namespace.yaml [hunk 6]
```diff
@@ -4,3 +4,4 @@ metadata:
   name: app
   labels:
     fleet.azure.com/name: test
+    region: east
```
[the remark this answers is not in the document]
Comment (britaniar) [reply pkg/controllers/rollout/manifests/test_namespace.yaml:7]: Looked into it and tested without it, looks like it is not needed.

File: pkg/controllers/rollout/suite_test.go (hunk 5 above)
Comment (zhiying-lin) [reply pkg/controllers/rollout/suite_test.go:109]: nope, we even did not create the webhook sever.

We need something like  https://github.com/Azure/fleet/blob/main/cmd/hubagent/main.go#L111 when creating the manager.

Comment (britaniar) [reply pkg/controllers/rollout/suite_test.go:109]: As discussed offline, webhook is being called in rollouts controller.go file which is why the integration test was failing previously. Need to revisit and discuss if needed here.

File: pkg/controllers/rollout/controller.go [hunk 7]
```diff
@@ -123,6 +123,7 @@ func (r *Reconciler) Reconcile(ctx context.Context, req runtime.Request) (runtim
 
 	// fill out all the default values for CRP just in case the mutation webhook is not enabled.
 	fleetv1beta1.SetDefaultsClusterResourcePlacement(&crp)
+	// TODO: revisit to see if we should remove logic for validation
```
Comment (zhiying-lin) [inline pkg/controllers/rollout/controller.go:126]: @ryanzhang-oss i'm thinking of removing this validation. The current validation is very heavy and we should assume validation webhook is enabled.

**Review by zhiying-lin — COMMENTED** [review]
please resolve the file conflicts.

File: pkg/controllers/rollout/controller.go (hunk 7 above)
Comment (ryanzhang-oss) [reply pkg/controllers/rollout/controller.go:126]: SGTM as the webhook is fail to close

File: pkg/controllers/rollout/suite_test.go (hunk 5 above)
Comment (zhiying-lin) [reply pkg/controllers/rollout/suite_test.go:109]: statement is not accurate, > "validation is being called", not the webhook

When you use the "webhook" terminology, it's talking about 
[k8 webhook](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#webhook-request-and-response)

**Review by zhiying-lin — APPROVED** [review]

**Pull request merged**